### PR TITLE
Allow existing "Python RAII guards" to be used as context managers

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #102643
* #102642
* __->__ #102579

This PR adds a `py_context_manager_DEPRECATED` that converts a C++ RAII
guard to an object that may be either used as Python context manager or
as a "Python RAII guard".

We don't convert all of them to Python context manager only due to BC
reasons; people in OSS and internally actually rely on these APIs and I
don't want to break them. We are justified in breaking BC if we wanted
to, but it seemed like too much work for not a lot of gain.

The API is postfixed with "DEPRECATED" to indicate that people should
really use `py_context_manager` (converts C++ RAII guard to Python
context manager) instead.

Test Plan:
- this PR converts all PyTorch usages of _AutoDispatchBelowAutograd to
context manager. I can do the rest in follow-ups.

### CODE REVIEW

File: torch/csrc/utils/python_raii.h [hunk 1]
```diff
@@ -42,5 +42,45 @@ void py_context_manager(const py::module& m, const char* name) {
...
+
+// Definition: a "Python RAII guard" is an object in Python that acquires
+// a resource on init and releases the resource on deletion.
+//
+// This API turns a C++ RAII guard into an object can be used either as a
+// Python context manager or as a "Python RAII guard".
+//
+// Please prefer `py_context_manager` to this API if you are binding a new
+// RAII guard into Python because "Python RAII guards" don't work as expected
+// in Python (Python makes no guarantees about when an object gets deleted)
```
Comment (bdhirsh) [inline torch/csrc/utils/python_raii.h:70]: thanks for explaining the footgun!

**Review by bdhirsh — APPROVED** [review]

**Review by albanD — APPROVED** [review]
SGTM!
Thanks for taking the time to fix this.
Any chance we could change back some of these to use the non-deprecated version in a follow up? Or all of these are actually used outside of our code?

**Comment by zou3519** [comment]
> Any chance we could change back some of these to use the non-deprecated version in a follow up? Or all of these are actually used outside of our code?

We can change most of these to use the non-deprecated version (and I'll do those in some follow ups). I won't touch the riskier ones (dispatchbelowautograd and the torch_dispatch/torch_function ones)

**Pull request closed without merging**